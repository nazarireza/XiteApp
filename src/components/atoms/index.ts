export * from './IconClose';
export * from './IconSearch';
export * from './IconBack';
export * from './Space';
export * from './IconButton';
export * from './IconFilter';
export * from './GenreItem';
export * from './VideoItem';
