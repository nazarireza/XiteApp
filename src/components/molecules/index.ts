export * from './SearchBox';
export * from './GenresList';
export * from './VideosList';
export * from './Header';
export * from './GenresWrapList';
export * from './YearsWrapList';
export * from './VideoCompactList';
