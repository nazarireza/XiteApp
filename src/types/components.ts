export type IconBaseProps = {
  color?: string;
  size?: number;
};

export type onListItemSelect<T> = (item: {
  item: T;
  index: number;
  selected?: boolean;
}) => void;
