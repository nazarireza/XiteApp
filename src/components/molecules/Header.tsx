import React, {memo} from 'react';
import {StyleSheet, View, TouchableOpacity, Text} from 'react-native';
import {colors, dictionary, typography} from '../../assets';
import {IconBack, Space} from '../atoms';

interface HeaderProps {
  title?: string;
  onBackPress?: () => void;
  actionTitle?: string;
  onActionPress?: () => void;
  dissmissable?: boolean;
}

export const Header: React.FC<HeaderProps> = memo(
  ({
    title = dictionary.filterModalTitle,
    actionTitle,
    onActionPress,
    onBackPress,
    dissmissable = true,
  }) => {
    const hasAction = !!actionTitle;
    return (
      <View style={styles.container}>
        <TouchableOpacity
          disabled={!dissmissable}
          style={styles.titleContainer}
          activeOpacity={0.7}
          onPress={onBackPress}>
          {dissmissable && (
            <>
              <IconBack color={colors.text.primary} />
              <Space size={8} />
            </>
          )}
          <Text style={typography.header}>{title}</Text>
        </TouchableOpacity>
        {hasAction && (
          <TouchableOpacity
            activeOpacity={0.7}
            style={styles.actionContainer}
            onPress={onActionPress}>
            <Text style={typography.headerAction}>{actionTitle}</Text>
          </TouchableOpacity>
        )}
      </View>
    );
  },
);

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
    paddingVertical: 8,
  },
  titleContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    flex: 1,
  },
  actionContainer: {
    alignItems: 'center',
    justifyContent: 'center',
  },
});
