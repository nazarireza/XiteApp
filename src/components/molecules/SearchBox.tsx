import React, {forwardRef, memo} from 'react';
import {StyleSheet, View, TextInput, TextInputProps} from 'react-native';
import {colors, dictionary} from '../../assets';
import {IconButton, IconClose, IconSearch, Space} from '../atoms';

interface SearchBoxProps {
  value?: string;
  placeHolder?: string;
  onChange?: (keyword: string) => void;
}

export const SearchBox = memo(
  forwardRef<TextInput, TextInputProps & SearchBoxProps>(
    (
      {placeHolder = dictionary.searchBoxPlaceHolder, onChange, value, ...rest},
      ref,
    ) => {
      const hasValue = !!value;
      return (
        <View style={styles.container}>
          <IconSearch color={colors.text.secondary} />
          <Space size={8} />
          <TextInput
            ref={ref}
            onChangeText={onChange}
            value={value}
            style={styles.input}
            placeholder={placeHolder}
            placeholderTextColor={colors.text.secondary}
            returnKeyType="search"
            {...rest}
          />
          {hasValue && (
            <IconButton onPress={() => onChange?.('')}>
              <IconClose />
            </IconButton>
          )}
        </View>
      );
    },
  ),
);

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
    backgroundColor: colors.background.card,
    paddingHorizontal: 16,
    borderRadius: 30,
  },
  input: {
    flex: 1,
    color: colors.text.primary,
    paddingVertical:12,
  },
});
