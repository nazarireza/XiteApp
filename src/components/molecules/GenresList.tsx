import React, {memo, useCallback} from 'react';
import {
  StyleSheet,
  FlatListProps,
  FlatList,
  ListRenderItem,
  View,
  TouchableOpacityProps,
} from 'react-native';
import {colors} from '../../assets';
import {Genre} from '../../types';
import {IconButton, IconFilter, Space, GenreItem, onListItemSelect} from '../atoms';

interface GenresListProps {
  onFilterPress?: () => void;
  selectedItems?: Array<Genre>;
  onItemSelect?: onListItemSelect<Genre>;
}

export const GenresList: React.FC<FlatListProps<Genre> & GenresListProps> =
  memo(({onFilterPress, selectedItems, onItemSelect, ...rest}) => {
    const renderItem: ListRenderItem<Genre> = useCallback(
      ({item, index}) => {
        const selected = selectedItems?.some(p => p.id == item.id);
        return (
          <GenreItem
            {...item}
            selected={selected}
            onPress={() => onItemSelect?.({item, index, selected: !selected})}
          />
        );
      },
      [selectedItems],
    );

    return (
      <View style={styles.container}>
        <FlatList
          ListHeaderComponent={() => <FilterButton onPress={onFilterPress} />}
          ItemSeparatorComponent={() => <Space size={8} />}
          showsHorizontalScrollIndicator={false}
          horizontal
          contentContainerStyle={styles.scrollContainer}
          {...rest}
          renderItem={renderItem}
        />
      </View>
    );
  });

const FilterButton: React.FC<TouchableOpacityProps> = memo(props => {
  return (
    <IconButton style={styles.filterButton} {...props}>
      <IconFilter />
    </IconButton>
  );
});

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
  },
  scrollContainer: {
    paddingHorizontal: 24,
    alignItems: 'center',
  },
  filterButton: {
    marginEnd: 8,
    backgroundColor: colors.primary,
    paddingVertical: 4,
    paddingHorizontal: 16,
    borderRadius: 5,
  },
});
