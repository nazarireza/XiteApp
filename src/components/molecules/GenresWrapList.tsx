import React, {memo, useCallback} from 'react';
import {Genre} from '../../types';
import {
  ToggleButton,
  WrapList,
  WrapListProps,
  WrapListRenderItem,
} from '../atoms';

export const GenresWrapList: React.FC<WrapListProps<Genre>> = memo(props => {
  const renderItem: WrapListRenderItem<Genre> = useCallback(
    ({item, index, selected, onSelect}) => {
      return (
        <ToggleButton
          title={item.name}
          selected={selected}
          onPress={() => onSelect?.({item, index, selected: !selected})}
        />
      );
    },
    [],
  );

  return (
    <WrapList
      {...props}
      {...{renderItem}}
      keyExtractor={(item: Genre) => `${item.id}`}
    />
  );
});
