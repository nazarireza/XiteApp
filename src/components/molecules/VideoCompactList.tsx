import React, {memo, useCallback} from 'react';
import {
  StyleSheet,
  FlatListProps,
  FlatList,
  ListRenderItem,
} from 'react-native';
import {NormalizedVideo} from '../../types';
import {Space, onListItemSelect, VideoCompactItem} from '../atoms';

interface VideosCompactListProps {
  onItemSelect?: onListItemSelect<NormalizedVideo>;
}

export const VideosCompactList: React.FC<
  FlatListProps<NormalizedVideo> & VideosCompactListProps
> = memo(({onItemSelect, ...rest}) => {
  const renderItem: ListRenderItem<NormalizedVideo> = useCallback(
    ({item, index}) => {
      return (
        <VideoCompactItem
          {...item}
          onPress={() => onItemSelect?.({item, index})}
        />
      );
    },
    [],
  );

  return (
    <FlatList
      keyExtractor={item => `${item.id}`}
      ItemSeparatorComponent={() => <Space />}
      contentContainerStyle={styles.scrollContainer}
      {...rest}
      renderItem={renderItem}
    />
  );
});

const styles = StyleSheet.create({
  scrollContainer: {},
});
