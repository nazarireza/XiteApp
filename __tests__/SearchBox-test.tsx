import 'react-native';
import React from 'react';
import {SearchBox} from '../src/components/molecules';
import renderer from 'react-test-renderer';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({adapter: new Adapter()});

it('renders correctly', () => {
  const wrapper = renderer.create(<SearchBox />).toJSON();
  expect(wrapper).toMatchSnapshot();
});

it('renders correctly with value', () => {
  const wrapper = renderer.create(<SearchBox value="value" />).toJSON();
  expect(wrapper).toMatchSnapshot();
});

it('renders correctly with placeholder', () => {
  const wrapper = renderer
    .create(<SearchBox placeHolder="placeholder" />)
    .toJSON();
  expect(wrapper).toMatchSnapshot();
});

it('onChange callback works correctly', () => {
  const onChange = jest.fn();
  const wrapper = shallow(<SearchBox onChange={onChange} />);

  wrapper.find({testID: 'input'}).simulate('changeText', 'query!');

  expect(onChange).toHaveBeenCalledWith('query!');
});
