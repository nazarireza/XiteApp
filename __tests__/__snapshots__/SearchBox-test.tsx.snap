// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#2E2E3F",
      "borderRadius": 30,
      "flexDirection": "row",
      "paddingHorizontal": 16,
    }
  }
>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={24}
    bbWidth={24}
    focusable={false}
    height={24}
    meetOrSlice={0}
    minX={0}
    minY={0}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "flex": 0,
          "height": 24,
          "width": 24,
        },
      ]
    }
    vbHeight={24}
    vbWidth={24}
    width={24}
  >
    <RNSVGGroup>
      <RNSVGPath
        d="M16.294 14.579h-.903l-.32-.31a7.4 7.4 0 001.795-4.836 7.433 7.433 0 10-7.433 7.433c1.84 0 3.533-.675 4.837-1.796l.309.32v.904L20.296 22 22 20.296l-5.706-5.717zm-6.861 0a5.139 5.139 0 01-5.146-5.146 5.138 5.138 0 015.146-5.146 5.139 5.139 0 015.146 5.146 5.14 5.14 0 01-5.146 5.146z"
        fill={4286152085}
        fillRule={1}
        propList={
          Array [
            "fill",
            "fillRule",
          ]
        }
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <View
    style={
      Object {
        "height": 8,
        "width": 8,
      }
    }
  />
  <TextInput
    placeholder="Artist, Title, ..."
    placeholderTextColor="#797D95"
    returnKeyType="search"
    style={
      Object {
        "color": "#F1F2F4",
        "flex": 1,
        "paddingVertical": 12,
      }
    }
    testID="input"
  />
</View>
`;

exports[`renders correctly with placeholder 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#2E2E3F",
      "borderRadius": 30,
      "flexDirection": "row",
      "paddingHorizontal": 16,
    }
  }
>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={24}
    bbWidth={24}
    focusable={false}
    height={24}
    meetOrSlice={0}
    minX={0}
    minY={0}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "flex": 0,
          "height": 24,
          "width": 24,
        },
      ]
    }
    vbHeight={24}
    vbWidth={24}
    width={24}
  >
    <RNSVGGroup>
      <RNSVGPath
        d="M16.294 14.579h-.903l-.32-.31a7.4 7.4 0 001.795-4.836 7.433 7.433 0 10-7.433 7.433c1.84 0 3.533-.675 4.837-1.796l.309.32v.904L20.296 22 22 20.296l-5.706-5.717zm-6.861 0a5.139 5.139 0 01-5.146-5.146 5.138 5.138 0 015.146-5.146 5.139 5.139 0 015.146 5.146 5.14 5.14 0 01-5.146 5.146z"
        fill={4286152085}
        fillRule={1}
        propList={
          Array [
            "fill",
            "fillRule",
          ]
        }
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <View
    style={
      Object {
        "height": 8,
        "width": 8,
      }
    }
  />
  <TextInput
    placeholder="placeholder"
    placeholderTextColor="#797D95"
    returnKeyType="search"
    style={
      Object {
        "color": "#F1F2F4",
        "flex": 1,
        "paddingVertical": 12,
      }
    }
    testID="input"
  />
</View>
`;

exports[`renders correctly with value 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#2E2E3F",
      "borderRadius": 30,
      "flexDirection": "row",
      "paddingHorizontal": 16,
    }
  }
>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={24}
    bbWidth={24}
    focusable={false}
    height={24}
    meetOrSlice={0}
    minX={0}
    minY={0}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "flex": 0,
          "height": 24,
          "width": 24,
        },
      ]
    }
    vbHeight={24}
    vbWidth={24}
    width={24}
  >
    <RNSVGGroup>
      <RNSVGPath
        d="M16.294 14.579h-.903l-.32-.31a7.4 7.4 0 001.795-4.836 7.433 7.433 0 10-7.433 7.433c1.84 0 3.533-.675 4.837-1.796l.309.32v.904L20.296 22 22 20.296l-5.706-5.717zm-6.861 0a5.139 5.139 0 01-5.146-5.146 5.138 5.138 0 015.146-5.146 5.139 5.139 0 015.146 5.146 5.14 5.14 0 01-5.146 5.146z"
        fill={4286152085}
        fillRule={1}
        propList={
          Array [
            "fill",
            "fillRule",
          ]
        }
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <View
    style={
      Object {
        "height": 8,
        "width": 8,
      }
    }
  />
  <TextInput
    placeholder="Artist, Title, ..."
    placeholderTextColor="#797D95"
    returnKeyType="search"
    style={
      Object {
        "color": "#F1F2F4",
        "flex": 1,
        "paddingVertical": 12,
      }
    }
    testID="input"
    value="value"
  />
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    nativeID="animatedComponent"
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "opacity": 1,
      }
    }
  >
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight={24}
      bbWidth={24}
      focusable={false}
      height={24}
      meetOrSlice={0}
      minX={0}
      minY={0}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          Object {
            "flex": 0,
            "height": 24,
            "width": 24,
          },
        ]
      }
      vbHeight={24}
      vbWidth={24}
      width={24}
    >
      <RNSVGGroup>
        <RNSVGPath
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          fill={4294046452}
          propList={
            Array [
              "fill",
            ]
          }
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
</View>
`;
